internal:Recover , detect the snapshot dir is exists

jokemanfire opened the pull request:
A simple way to reproduce, deleting the snapshots file of the container, containerd cannot self recover.
![image](https://github.com/user-attachments/assets/189c4164-eb09-4957-a7b9-def9dd78733f)

fuweid commented on internal/cri/server/images/snapshots.go:
```diff
@@ -109,6 +109,11 @@ func (s *snapshotsSyncer) sync() error {
 				if !errdefs.IsNotFound(err) {
+					// delete the snapshot stats if it's not found, this can happen
+					// when the snapshot fs is removed.
+					if err := snapshotter.Remove(ctx, info.Name); err != nil {
```
I don't think we should do that. it's read-only operation. why do we need to handle deletion in this loop?
If power outage happens, host restarts and all the containers exited. kubelet will delete them. and then snapshot will be deleted as well. if you're using other container managerment rather than kubelet, please handle it in managerment side. it makes more senses.

dmcgowan replied:
Agreed, we don't attempt to clean up from disk corruption but can try to error loudly when its seen. Likely the safest thing to do after corruption is to restore the node to a previous state or at least remove and re-pull containers.

jokemanfire replied:
At that time, Cli was unable to solve the problem in the production environment, so we tried to solve it from this side. I will take another look at this issue.

jokemanfire replied:
I found that this error message is transparent to the upper level management tool, whether it is restarting containerd or restarting the computer (PS:/run directory is not using tmpfs). The cri tool will not find the existence of the container, but the database still has a record of this snapshot

jokemanfire replied:
It is . The situation is that when containerd is running normally, it suddenly loses power and restarts, causing the cri interface to query that there is no container, but this snapshot record exists in the database. The test was conducted using `Bucketbench `after my modification (for concurrent testing), and I will try to reproduce it in white box. Black box state is indeed difficult to reproduce stably.

jokemanfire committed:
internal:In case of abnormal power outage. Causing snapshots to remain in the database.

Try to change status of container's status, when the snapshot file have been deleted

Signed-off-by: jokemanfire <hu.dingyang@zte.com.cn>

jokemanfire commented:
@fuweid  @dmcgowan  If the snapshot file is lose ,the container and sandbox's status will be changed in Cri, Avoid using container mirroring operations with incorrect states. I change it in recover

fuweid commented:
> If the snapshot file is lose ,the container and sandbox's status will be changed in Cri, Avoid using container mirroring operations with incorrect states. I change it in recover

Sorry, I still don't get your point. 

If power outage causes unexpected data lost, why doesn't kubelet delete containers created before power outage?
Just want to understand it more before reviewing your code. If you doesn't use kubelet to manage container, please let us know. CRI is designed for kubelet instead of common container manager. If you build your own container manager component, containerd API is good choice instead of CRI. 

IMO, I just want to narrow down the scope of CRI plugin.


Thanks.

jokemanfire commented:
> > If the snapshot file is lose ,the container and sandbox's status will be changed in Cri, Avoid using container mirroring operations with incorrect states. I change it in recover
> 
> Sorry, I still don't get your point.
> 
> If power outage causes unexpected data lost, why doesn't kubelet delete containers created before power outage? Just want to understand it more before reviewing your code. If you doesn't use kubelet to manage container, please let us know. CRI is designed for kubelet instead of common container manager. If you build your own container manager component, containerd API is good choice instead of CRI.
> 
> IMO, I just want to narrow down the scope of CRI plugin.
> 
> Thanks.

I will describe the problem more detailed.
1. power outage which we can't control , it is an abnormal power outage that the controller is unaware of beforehand.We use crictl and nerdctl more often.
2. In scene 1, find the container in db has been removed , but snapshots still in db ,and snapshots dir has been removed.
3.  When attempting to reproduce this problem , I also found containerd's cri interface will not monitor whether the snapshot has been deleted(Just try to remove snapshots's dir), the container still in running status (which this pr try to resolve)
4. In scene 2, I think the snapshot's residues, which can also be completed during the recover phase, but there may be a lot of modified code and we are not planning to address it in this PR.
5. What is the current goal of the community for the CRI interface? (Should it be abandoned?) Because we use it more often.

fuweid commented:
> In scene 1, find the container in db has been removed , but snapshots still in db ,and snapshots dir has been removed.
When attempting to reproduce this problem , I also found containerd's cri interface will not monitor whether the snapshot has been deleted(Just try to remove snapshots's dir), the container still in running status (which this pr try to resolve)
In scene 2, I think the snapshot's residues, which can also be completed during the recover phase, but there may be a lot of modified code and we are not planning to address it in this PR.

If power outage causes data loss, my suggestion is that admin should get involved to check what impact there is. Based on your description, it's impossible to recover all the cases, for example, you can delete any files related to containerd or just delete one file in snapshot dir. It's unlikely to recover it from that situation. So, this change doesn't look good to me.

For user perspective, it will be great to have tool to detect any data loss or inconsistent issue. Once tool detects there is something wrong, cluster admin can cordon and drain this node to prevent from running into further nexpected issues.

> What is the current goal of the community for the CRI interface? (Should it be abandoned?) Because we use it more often.

CRI is only designed to k8s. It's not for non-k8s components.

jokemanfire commented:
> If power outage causes data loss, my suggestion is that admin should get involved to check what impact there is. Based on your description, it's impossible to recover all the cases, for example, you can delete any files related to containerd or just delete one file in snapshot dir. It's unlikely to recover it from that situation. So, this change doesn't look good to me.
> 
> For user perspective, it will be great to have tool to detect any data loss or inconsistent issue. Once tool detects there is something wrong, cluster admin can cordon and drain this node to prevent from running into further nexpected issues.

Forgive me for not expressing myself properly. In fact, I did not do any recovery, but only made the state aware by the cri interface (I set the state to unknown). The snapshots were lost so that the upper level caller could know and take action.

jokemanfire commented:
> For user perspective, it will be great to have tool to detect any data loss or inconsistent issue. Once tool detects there is something wrong, cluster admin can cordon and drain this node to prevent from running into further nexpected issues.

Is the community doing this? I think it would be great if there were a detection containerd service that includes a database and a detection tool for data stored on disk !

fuweid commented:
Hi @jokemanfire sorry for late reply.

> Forgive me for not expressing myself properly. In fact, I did not do any recovery, but only made the state aware by the cri interface (I set the state to unknown). The snapshots were lost so that the upper level caller could know and take action.

If I understand it correctly, you're saying that for the running containers, some operations delete the snapshot and then restart containerd. And then containerd will show a lot of log about not-found error. If so, I would like to keep it as it is. The error can be throwed to cluster administrator and then administrators detect that issue as soon as possible.

> NOTE: By default, the snapshot is referenced by container. Even if user uses `ctr snapshot delete`, the GC component will keep it since that owner, container, still exists in db. If snapshot db (boltdb file) is corrupted from unexpected kernel issue or power outage, all the containers should be in exited status and kubelet will delete all exited containers. So, user won't see this error in log file.

> Is the community doing this? I think it would be great if there were a detection containerd service that includes a database and a detection tool for data stored on disk !

It's just my thought :) Feel free to file feature-request in `ctr` for that.

jokemanfire commented:
> The error can be throwed to cluster administrator and then administrators detect that issue as soon as possible.

I think this does not require manual intervention. It only needs to detect that the snapshot does not exist during the recover phase, and then set the status to non running. The upper layer, such as k8s, will perceive it and recreate the container. (Even if the upper layer is not k8s, as long as the queried status is incorrect, the upper layer caller will know that the container is abnormal). If only printing information, it is indeed possible to locate it manually, but I think it is a bit wasteful because it is easier to inform the problem on the containerd side.

fuweid commented:
> I think this does not require manual intervention. It only needs to detect that the snapshot does not exist during the recover phase, and then set the status to non running.

That's why I want to know how to reproduce this issue. The snapshot is referenced by container so it's unlikely to have a case that container exits but snapshot is gone. As a reviewer, I would like to know the details before adding logic. Based on your description, I still insist to keep it as it is.

Wei
